src/main/java/org/itsallcode/junit/sysextensions/SystemOutGuard.java

```diff
@@ -0,0 +1,146 @@
+ * 
+ * You should have received a copy of the GNU General Public
+ * License along with this program.  If not, see
+ * <http://www.gnu.org/licenses/gpl-3.0.html>.
+ * #L%
+ */
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.PrintStream;
+import java.lang.reflect.Executable;
+
+import org.itsallcode.io.CapturingOutputStream;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.AfterAllCallback;
+import org.junit.jupiter.api.extension.BeforeAllCallback;
+import org.junit.jupiter.api.extension.BeforeEachCallback;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.ExtensionContext.Namespace;
+import org.junit.jupiter.api.extension.ParameterContext;
+import org.junit.jupiter.api.extension.ParameterResolutionException;
+import org.junit.jupiter.api.extension.ParameterResolver;
+
+/**
+ * This class implements a JUnit extension that lets you capture data written to
+ * {@link System#out}. in
+ *
+ * <p>
+ * For this it provides the user with a special kind of {@link OutputStream},
+ * the {@link CapturingOutputStream}. If you add that stream as a parameter to a
+ * method annotated with either <code>@BeforeTest</code> or <code>@Test</code> ,
+ * the {@link SystemOutGuard} injects an instance of the
+ * {@link CapturingOutputStream}.
+ * </p>
+ *
+ * <p>
+ * For the users convenience capturing is automatically reset before each call
+ * to the method annotated with <code>@Before</code>.
+ */
+public class SystemOutGuard implements BeforeAllCallback, BeforeEachCallback, ParameterResolver, AfterAllCallback
+{
+    private static final String PREVIOUS_OUTPUT_STREAM_KEY = "PREV_OSTREAM";
+    private static final String CAPTURING_OUTPUT_STREAM_KEY = "CAPT_OSTREAM";
+
+    @Override
+    public void beforeAll(final ExtensionContext context) throws Exception
+    {
+        saveCurrentSystemOut(context);
+        flushSystemOut();
+        replaceSystemOutWithCapturingStream(context);
+
+    }
+
+    private void saveCurrentSystemOut(final ExtensionContext context)
+    {
+        context.getStore(getNamespace()).put(PREVIOUS_OUTPUT_STREAM_KEY, System.out);
+    }
+
+    private void flushSystemOut()
+    {
+        System.out.flush();
+    }
+
+    private void replaceSystemOutWithCapturingStream(final ExtensionContext context)
+    {
+        final CapturingOutputStream capturingStream = new CapturingOutputStream(System.out);
+        context.getStore(getNamespace()).put(CAPTURING_OUTPUT_STREAM_KEY, capturingStream);
+        System.setOut(new PrintStream(getCapturingOutputStream(context)));
+    }
+
+    private CapturingOutputStream getCapturingOutputStream(final ExtensionContext context)
+    {
+        return (CapturingOutputStream) context.getStore(getNamespace()).get(CAPTURING_OUTPUT_STREAM_KEY);
+    }
+
+    private Namespace getNamespace()
+    {
+        return Namespace.create(SystemOutGuard.class);
```

Comment:
Could be a constant instead of creating it again every time

Reply:
Good idea.